### PR TITLE
python@2: zlib and gdbm fixes

### PR DESCRIPTION
**Pull request by m000**
Tests for zlib and gdbm were failing on Mojave when building from source using the CommandLineTools SDK. zlib not working renders the `python@2` installation unusable, as the module is used by pip. Details in commit message.

-----

- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/master/CONTRIBUTING.md)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Does your build pass `brew audit --strict <formula>` (after doing `brew install <formula>`)?

-----

### CODE REVIEW

**Comment by fxcoudert** [comment]
Depending on Homebrew zlib while there is a perfectly fine system zlib in macOS is not an option.

**Comment by m000** [comment]
@fxcoudert I figured out that zlib works when compiling with Xcode SDK. But when compiling with CommandLineTools SDK, it is ignored. Any hints on why that would happen? Maybe this occurs in other packages too.

**Comment by fxcoudert** [comment]
We are not going to hack `setup.py` in Homebrew. If this is a proper fix, please submit it to Python, so everyone can benefit from it.

**Comment by m000** [comment]
@fxcoudert This is a workaround for homebrew. I've spent a fair amount of time debugging the issue. If anyone else is willing to pick up from here and do the required testing for upstream python, I'm fine with that. 

setup.py is already being patched by the recipe to get sqlite3 and readline to work. I don't see why this fix is a problem.

**Comment by SMillerDev** [comment]
@brewtestbot test this please!

**Comment by m000** [comment]
@SMillerDev Maintainer feedback never came. Was there anything amiss?

**Comment by zbeekman** [comment]
> setup.py is already being patched by the recipe to get sqlite3 and readline to work. I don't see why this fix is a problem.

Generally we don't accept patches unless at least a bug report is submitted upstream, if not a proper patch.

**Comment by zbeekman** [comment]
Can folks from @Homebrew/core with more python experience weigh in here and on https://github.com/Homebrew/homebrew-core/pull/38268?

**Comment by sjackman** [comment]
Homebrew/linuxbrew-core has a similar patch, so it seems like it is a problem.
See https://github.com/Homebrew/linuxbrew-core/blob/a0d5dc7bcafc9681aa327d404d5c3f633f8b5e0d/Formula/python%402.rb#L110-L121

**Comment by zbeekman** [comment]
@BrewTestBot test this please

**Comment by iMichka** [comment]
I am ok to add the patch for the cppflags, ldflags, as we have the same on Linux, and it does not harm to have it on mac.

For the setup.py change, this should be submitted to upstream first.

Also, you have just the CLT installed I guess. The docs (https://docs.brew.sh/Installation) tell the following:

> Most formulae require a compiler. A handful require a full Xcode installation. 

Would this formula fall in the category "a handful" ?

**Comment by m000** [comment]
@iMichka I will try to upstream the change sure.

But note that python@2 ranks [#9 in number of installations](https://formulae.brew.sh/analytics/install/365d/) for the past year. So, this is a *very* popular recipe, and waiting for the upstream to make a new release would probably inconvenience a sizeable chunk of users for a fair amount of time.

So I think that it's best for the homebrew community to also independently fix the issue here, while I pursue an upstream fix. Since I'll be following the upstream patch too, I am willing to a new PR that removes the fix from the formula when setup.py is fixed in a released version.

Yes, I have only the CLT installed.

IMHO, due to its popularity, if python@2 can be built only with CLT, then it should. Dragging Xcode along for something that a lot of people use homebrew for is like introducing an implicit dependency for running homebrew.

**Comment by MikeMcQuaid** [comment]
> Would this formula fall in the category "a handful" ?

Yes

> So, this is a _very_ popular recipe, and waiting for the upstream to make a new release would probably inconvenience a sizeable chunk of users for a fair amount of time.

Similarly, this is a very popular recipe so we need to be extremely careful making changes that upstream haven't (yet) approved due to the risk in breaking things unexpectedly. This is why we'd like to see the patch at least submitted upstream (ideally reviewed and merged although not necessary to be released) before we ship it to all Homebrew users.

> IMHO, due to its popularity, if python@2 can be built only with CLT, then it should. Dragging Xcode along for something that a lot of people use homebrew for is like introducing an implicit dependency for running homebrew.

We agree here.

**Comment by zbeekman** [comment]
OK, someone else can correct me if I'm not accurately representing the path forward from Homebrew's perspective, but, @m000 can you please do the following:

1. Link to the upstream Issue or Pull Request you've created in a comment/response here
2. Change this PR to replace, at a minimum, the `inreplace` block with a proper patch, that was submitted upstream
3. Open a PR at https://github.com/Homebrew/formula-patches named something like `python@2/2.7.16-zlib-bug.patch`

@fxcoudert @MikeMcQuaid Do we want to depend on a full XCode install or a brewed zlib?

File: Formula/python@2.rb
Comment (zbeekman) [inline Formula/python@2.rb]: @m000 is this for `gdbm` only or is this required for finding other dependencies too when homebrew is not in `/usr/local`?

**Comment by MikeMcQuaid** [comment]
> @fxcoudert @MikeMcQuaid Do we want to depend on a full XCode install or a brewed zlib?

A brewed zlib would be ideal but if we're going to be patching then using the system zlib would be preferable.

File: Formula/python@2.rb
Comment (zbeekman) [inline Formula/python@2.rb]: This should be submitted in patch form, to match what is happening upstream, with a comment linking to the issue or PR upstream.

From your comment, it sounds like a fix is needed in the `is_macosx_sdk_path()` method rather than in `setup.py`. Is that true? Any idea how to patch that?

**Review by zbeekman — CHANGES_REQUESTED** [review]

**Comment by zbeekman** [comment]
> A brewed zlib would be ideal but if we're going to be patching then using the system zlib would be preferable.

Sorry, I didn't mean to edit Mike's comment above... need some more coffee apparently. I returned it to its original state.

Here is what I was trying to say in my own comment:

I think I was misunderstanding the conversation above. To me, it appears that there is a defect upstream in how they're handling sdk_path for finding zlib, so they are not finding it correctly. IIUC, then the proposed patching of `setup.py` would fix this so that the system zlib could be found. @m000 if my understanding is erroneous, please correct me.

Thanks!

**Comment by fxcoudert** [comment]
Closing due to lack of activity, this needs to be understood more before it can be fixed.

**Pull request closed without merging**
